### PR TITLE
Port to FreeBSD

### PR DESCRIPTION
**Pull request by kev009**
This is a fairly rough pass at porting omr to FreeBSD which I did as a way to gauge how much work it would be.  Now that I understand the breadth, I am submitting this for early feedback on the viability and any pointers for finishing the port such as test documentation or notes from a previous port.

### CODE REVIEW

File: fvtest/compilertest/build/platform/host/amd64-freebsd-clang.mk [hunk 1]
```diff
@@ -0,0 +1,27 @@
+###############################################################################
+# Copyright (c) 2016, 2017 IBM Corp. and others
```
Comment (Leonardo2718) [inline fvtest/compilertest/build/platform/host/amd64-freebsd-clang.mk]: The copyright header should have the current year: `Copyright (c) 2016, 2018 IBM Corp. and others`.

File: jitbuilder/build/platform/host/amd64-freebsd-clang.mk [hunk 2]
```diff
@@ -0,0 +1,28 @@
+###############################################################################
+# Copyright (c) 2016, 2016 IBM Corp. and others
```
Comment (Leonardo2718) [inline jitbuilder/build/platform/host/amd64-freebsd-clang.mk]: The copyright header should have the current year: `Copyright (c) 2016, 2018 IBM Corp. and others`.

File: port/freebsd/omrosdump_helpers.c [hunk 3]
```diff
@@ -0,0 +1,169 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2016 IBM Corp. and others
```
Comment (Leonardo2718) [inline port/freebsd/omrosdump_helpers.c]: The copyright header should have the current year: `Copyright (c) 2016, 2018 IBM Corp. and others`.

File: port/freebsd/omrosdump_helpers.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2016 IBM Corp. and others
```
Comment (Leonardo2718) [inline port/freebsd/omrosdump_helpers.h]: The copyright header should have the current year: `Copyright (c) 2016, 2018 IBM Corp. and others`.

File: port/freebsdamd64/omrsignal_context.c [hunk 5]
```diff
@@ -0,0 +1,323 @@
+/*******************************************************************************
+ * Copyright (c) 1991, 2015 IBM Corp. and others
```
Comment (Leonardo2718) [inline port/freebsdamd64/omrsignal_context.c]: The copyright header should have the current year: `Copyright (c) 1991, 2018 IBM Corp. and others`.

File: port/freebsdamd64/omrsignal_context.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2016 IBM Corp. and others
```
Comment (Leonardo2718) [inline port/freebsdamd64/omrsignal_context.h]: The copyright header should have the current year: `Copyright (c) 2016, 2018 IBM Corp. and others`.

File: thread/osx/priority.c [hunk 7]
```diff
@@ -49,47 +45,43 @@ initialize_priority_map(void)
...
+static int getCurrentThreadParams(int *prio, int *policy);
+
+/* regular priorities and policies: those available to java.lang.Threads,
+ */
+
 static int policy_regular_thread;
 static int policy_realtime_thread;
 
-/* Represents the O/S limits for realtime priorities */
+/* represents the O/S limits for realtime priorities */
```
Comment (Leonardo2718) [inline thread/osx/priority.c]: Please avoid mixing these kinds of stylistic changes with functional changes as they add "noise" and making review changes more difficult.

File: fvtest/compilertest/build/platform/host/amd64-freebsd-clang.mk [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+# [1] https://www.gnu.org/software/classpath/license.html
+# [2] http://openjdk.java.net/legal/assembly-exception.html
+#
+# SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+###############################################################################
+
+HOST_ARCH=x
+HOST_SUBARCH=amd64
+HOST_BITS=64
+OS=osx
```
Comment (jbeich) [inline fvtest/compilertest/build/platform/host/amd64-freebsd-clang.mk]: Typo?

File: fvtest/sigtest/sigTest.cpp [hunk 9]
```diff
@@ -69,7 +69,13 @@ enum TestAction {
 #endif /* defined(OMR_OS_WINDOWS) || defined(J9ZOS390) */
 #endif
 
+#if defined(FREEBSD)
+static sigjmp_buf env;
+static jmp_buf jenv;
+#else /* defined(FREEBSD) */
 static jmp_buf env;
+static jmp_buf jenv;
+#endif /* defined(FREEBSD) */
```
Comment (ymanton) [inline fvtest/sigtest/sigTest.cpp]: Did you intend to push this? Looks like this isn't necessary and would have been fine the way it was before based on your usage below.

Comment (kev009) [reply fvtest/sigtest/sigTest.cpp]: Yes see https://www.freebsd.org/cgi/man.cgi?query=setjmp&sektion=3 "Pairs of calls may	be intermixed, i.e., both sigsetjmp() and siglongjmp() and setjmp() and longjmp() combinations may be used in the same program, however, individual calls may not,	e.g. the env argument to setjmp() may not be passed to siglongjmp()."

It appears this is not explicitly defined on Linux but I didn't check the implementation http://man7.org/linux/man-pages/man3/setjmp.3.html

File: thread/osx/priority.c [hunk 10]
```diff
@@ -405,7 +413,7 @@ set_priority_spread(void)
 	if (dups < 0) {
 		dups = 0;
 	}
-	for (i = J9THREAD_PRIORITY_NORMAL - dups - 1, j = prio; i >= J9THREAD_PRIORITY_USER_MIN; i--) {
+	for (i = J9THREAD_PRIORITY_NORMAL - dups - 1, j = prio; i >= J9THREAD_PRIORITY_USER_MIN; i--, j) {
```
[the remark this answers is not in the document]
Comment (kev009) [reply thread/osx/priority.c]: Oops, all these changes to osx priority.c were accidental this looks like I copied the Linux one into OSX.  I waffled back and forth various times and with various files between using Linux and OSX as the port base and probably swapped 'diff' for a 'cp' in the wrong direction here.

**Comment by kev009** [comment]
@mstoodle my main intent is to bring up OpenJ9 for FreeBSD on POWER9.  I'm not sure whether that will be ppc64 or ppc64le right yet (we are big endian for now still) but amd64 is a useful stepping stone and usable in its own right.

I will amend this PR with suggested changes and leave it up for a couple days for general feedback then will probably close it after spending some time figuring out how to submit smaller mergable chunks per your feedback.

File: fvtest/sigtest/sigTest.cpp (hunk 9 above)
Comment (ymanton) [reply fvtest/sigtest/sigTest.cpp]: You're right, I didn't look at the rest of the file and didn't realize the existing Linux code was using both and passing a `jmp_buf` to `sigsetjmp`. The Linux version was and still is wrong isn't it? Shouldn't the BSD side of this `#ifdef` be used by both?

Comment (kev009) [reply fvtest/sigtest/sigTest.cpp]: @ymanton ok I can probably just remove the ifdef and have two trackers then

File: port/freebsdamd64/omrsignal_context.c [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+									"xmm11",
+									"xmm12",
+									"xmm13",
+									"xmm14",
+									"xmm15"
+								  };
+
+	if ((index >= 0) && (index < NXMMREGS)) {
+		*name = n_xmm[index];
+		*value = (void *) &floatState->sv_xmm[index];
```
Comment (kev009) [inline port/freebsdamd64/omrsignal_context.c]: It was a very wild guess as to whether this and L103 are correct

File: port/freebsdamd64/omrsignal_context.c [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+		*value = &context->sc_err;
+		return OMRPORT_SIG_VALUE_ADDRESS;
+	case 8:
+		*name = "TRAPNO";
+		*value = &context->sc_trapno;
+		return OMRPORT_SIG_VALUE_ADDRESS;
+	case 9:
+		*name = "OLDMASK";
+		abort();
+		/* XXX Linux?? *value = &context->oldmask; */
```
Comment (kev009) [inline port/freebsdamd64/omrsignal_context.c]: Couldn't easily figure out what this did in glibc on Linux

File: port/freebsdamd64/omrsignal_context.c [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+		return OMRPORT_SIG_VALUE_ADDRESS;
+	case 10:
+		*name = "CR2";
+		abort();
+		/* XXX Linux?? *value = &context->cr2; */
```
Comment (kev009) [inline port/freebsdamd64/omrsignal_context.c]: Couldn't easily figure out what this did in glibc on Linux

**Comment by kev009** [comment]
@mstoodle is there a doc anywhere that explains how the debug tracing stuff works?  I've gotten this pretty far along now into fvtest and need to chase out the cause of the remaining test failures.

**Comment by charliegracie** [comment]
@kev009 one thing to note is that the signal tests are failing in the Linux x86 and OS X builds with your changes.  You can see the failures here:
https://travis-ci.org/eclipse/omr/jobs/423986644

**Comment by dnakamura** [comment]
One piece of feedback I would give, its that while we are going in and modifying the checks, it probably makes sense to conditionalize based on tested behavior rather than basing on OS. (which is something I would like to see in the codebase regardless, but adding a new platform is a good excuse to actually do it)

example:
instead of 
```
#if defined(LINUX) || defined(OSX) || defined(FREEBSD)
#include <unistd.h>
#endif
```
we instead do something like 
```
#if defined(PLATFORM_REQUIRES_UNISTD_H)
#include <unistd.h>
#endif
```
 where  `PLATFORM_REQUIRES_UNISTD_H` is determined based on autoconf/cmake checks

**Comment by kev009** [comment]
@dnakamura thanks that is good advice!  I'm going to try and get fvtest mostly working and then will take that effort up in my incremental PRs (this one I am considering a proof of concept)

**Comment by dnakamura** [comment]
also just to clarify: I dont expect you to rewrite all the platform specific checks, thats obviously a bunch of work. I was just suggesting fixing up the checks which would affect freeBSD

**Comment by Leonardo2718** [comment]
@kev009 any updates on this?

**Comment by kev009** [comment]
@Leonardo2718 I just rebased to head and added cmake support, and am working through fvtest failures

**Comment by kev009** [comment]
@dnakamura 
> where `PLATFORM_REQUIRES_UNISTD_H` is determined based on autoconf/cmake checks

Can you point to or create an example of this in autoconf and cmake for me to follow?

**Comment by charliegracie** [comment]
ping @dnakamura

**Comment by dnakamura** [comment]
@kev009 I was thining about it further, and It would be non trivial to figure out if unistd.h is actually needed for whatever it is we are doing. The easier approach is to just include unistd.h if we have it. (worst case we slightly increase compile time unnecessarily). I can do up an example on how to check for this, but there should be plenty of examples on how to check if a header file exists

**Comment by kev009** [comment]
@dnakamura I was hoping for a more complete analysis of where to implement feature based patterns rather than OS detection, not limited to the include file you originally mentioned this about

**Comment by dnakamura** [comment]
@kev009  We currently dont have any feature based configuration stuff in cmake, and as such we havent really decided on where it would go. I am open to suggestions.

**Comment by kev009** [comment]
rebased to head as of right now.  still some undiagnosed hangs/errors in ctest

**Comment by 0xdaryl** [comment]
There has not been any activity on this PR in nearly two years.  If there are no immediate plans to move it forward, I suggest we close it until it is ready to contribute.

**Pull request closed without merging**